# Exploratory Data Analysis
the goal of this Notebook to show Data Handling skills dealing with Missing Values and working with libraries such as Pandas and Numpy

In [199]:
import pandas as pd
import numpy as np

In [200]:
filepath = '/content/Unicorn_Companies.csv'
data = pd.read_csv(filepath)

In [201]:
data.head()

Company Valuation ($B) Date Joined        Country           City  \
0   Bytedance           $140    4/7/2017          China        Beijing   
1      SpaceX         $100.3   12/1/2012  United States      Hawthorne   
2      Stripe            $95   1/23/2014  United States  San Francisco   
3      Klarna          $45.6  12/12/2011         Sweden      Stockholm   
4  Epic Games            $42  10/26/2018  United States           Cary   

                  Industry                                  Select Inverstors  \
0  Artificial intelligence  Sequoia Capital China, SIG Asia Investments, S...   
1                    Other  Founders Fund, Draper Fisher Jurvetson, Rothen...   
2                  Fintech        Khosla Ventures, LowercaseCapital, capitalG   
3                  Fintech  Institutional Venture Partners, Sequoia Capita...   
4                    Other              Tencent Holdings, KKR, Smash Ventures   

  Founded Year Total Raised Financial Stage Investors Count Deal Terms  \
0         2012       $7.44B             IPO              28          8   
1         2002      $6.874B            None              29         12   
2         2010      $2.901B           Asset              39         12   
3         2005      $3.472B        Acquired              56         13   
4         1991      $4.377B        Acquired              25          5   

  Portfolio Exits  
0               5  
1            None  
2               1  
3               1  
4               2

In [202]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1037 entries, 0 to 1036
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Company            1037 non-null   object
 1   Valuation ($B)     1037 non-null   object
 2   Date Joined        1037 non-null   object
 3   Country            1037 non-null   object
 4   City               1037 non-null   object
 5   Industry           1037 non-null   object
 6   Select Inverstors  1037 non-null   object
 7   Founded Year       1037 non-null   object
 8   Total Raised       1037 non-null   object
 9   Financial Stage    1037 non-null   object
 10  Investors Count    1037 non-null   object
 11  Deal Terms         1037 non-null   object
 12  Portfolio Exits    1037 non-null   object
dtypes: object(13)
memory usage: 105.4+ KB


After having at the initial look at the data, i can see that we have no NULL entries but there is custom NULL entries like the case of data['Financial Stage'][1] it's written as NONE in String type, i want to customize the dataset so that it reads the None of type object and transfer it to np.nan so that we cal deal with it according to the use case

In [203]:
na_value = ['None']
df = pd.read_csv(filepath, na_values = na_value)
# it will tread any values None as np.nan

## in this dataset i can notice that missing data are written as NA, Missing,or None

In [204]:
df.info(), df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1037 entries, 0 to 1036
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Company            1037 non-null   object 
 1   Valuation ($B)     1037 non-null   object 
 2   Date Joined        1037 non-null   object 
 3   Country            1037 non-null   object 
 4   City               1037 non-null   object 
 5   Industry           1037 non-null   object 
 6   Select Inverstors  1020 non-null   object 
 7   Founded Year       994 non-null    float64
 8   Total Raised       1013 non-null   object 
 9   Financial Stage    49 non-null     object 
 10  Investors Count    1036 non-null   float64
 11  Deal Terms         1008 non-null   float64
 12  Portfolio Exits    49 non-null     float64
dtypes: float64(4), object(9)
memory usage: 105.4+ KB


(None, Company                0
 Valuation ($B)         0
 Date Joined            0
 Country                0
 City                   0
 Industry               0
 Select Inverstors     17
 Founded Year          43
 Total Raised          24
 Financial Stage      988
 Investors Count        1
 Deal Terms            29
 Portfolio Exits      988
 dtype: int64)

now we can see that we have many NULL objects in our dataset.

In [205]:
#we can easily check what different values we have in a column by using the .unique() method 
m = pd.DataFrame(data['Portfolio Exits'].unique())
m

0
0     5
1  None
2     1
3     2
4     3

In [206]:
 print(f"type of df['Financial Stage'][1] = {type(df['Financial Stage'][1])}") 
 # we can n now see the data['Financial Stage][1] = np.nan and we will see it as a float

type of df['Financial Stage'][1] = <class 'float'>


## This dataset has 1037 rows with 13 columns,, and all columns has dtype object, expect for Founded Year, Investors Count, Deal Terms, and Portfolio Exits  has type floats
- first step we need to convert columns (numbers in Valuation (```$```B), and Total Raised) to the real number, EX:insted of  ```$```140 B i want to have 140000000000
- Second step removing the ```$``` sign from the title, Valuation (```$```B) will become Valuation (```$```)

In [207]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1037 entries, 0 to 1036
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Company            1037 non-null   object 
 1   Valuation ($B)     1037 non-null   object 
 2   Date Joined        1037 non-null   object 
 3   Country            1037 non-null   object 
 4   City               1037 non-null   object 
 5   Industry           1037 non-null   object 
 6   Select Inverstors  1020 non-null   object 
 7   Founded Year       994 non-null    float64
 8   Total Raised       1013 non-null   object 
 9   Financial Stage    49 non-null     object 
 10  Investors Count    1036 non-null   float64
 11  Deal Terms         1008 non-null   float64
 12  Portfolio Exits    49 non-null     float64
dtypes: float64(4), object(9)
memory usage: 105.4+ KB


In [208]:
df['Founded Year'] 

0       2012.0
1       2002.0
2       2010.0
3       2005.0
4       1991.0
         ...  
1032    2015.0
1033    2019.0
1034    2011.0
1035    2017.0
1036    2008.0
Name: Founded Year, Length: 1037, dtype: float64

In [209]:
#having a copy to make to compare after transforming
df_copy = df.copy()

In [210]:
#cahning the name of our columns using the .rename() method
df_copy.rename(columns = {'Valuation ($B)':'Valuation ($)'},inplace=True)
df_copy.rename(columns = {'Total Raised':'Total Raised ($)'},inplace=True)
#having a list of all our columns names will be usfull 
col_names = df_copy.columns.tolist()
col_names

['Company',
 'Valuation ($)',
 'Date Joined',
 'Country',
 'City',
 'Industry',
 'Select Inverstors',
 'Founded Year',
 'Total Raised ($)',
 'Financial Stage',
 'Investors Count',
 'Deal Terms',
 'Portfolio Exits']

In [211]:
#checking 
df_copy.head()

Company Valuation ($) Date Joined        Country           City  \
0   Bytedance          $140    4/7/2017          China        Beijing   
1      SpaceX        $100.3   12/1/2012  United States      Hawthorne   
2      Stripe           $95   1/23/2014  United States  San Francisco   
3      Klarna         $45.6  12/12/2011         Sweden      Stockholm   
4  Epic Games           $42  10/26/2018  United States           Cary   

                  Industry                                  Select Inverstors  \
0  Artificial intelligence  Sequoia Capital China, SIG Asia Investments, S...   
1                    Other  Founders Fund, Draper Fisher Jurvetson, Rothen...   
2                  Fintech        Khosla Ventures, LowercaseCapital, capitalG   
3                  Fintech  Institutional Venture Partners, Sequoia Capita...   
4                    Other              Tencent Holdings, KKR, Smash Ventures   

   Founded Year Total Raised ($) Financial Stage  Investors Count  Deal Terms  \
0        2012.0           $7.44B             IPO             28.0         8.0   
1        2002.0          $6.874B             NaN             29.0        12.0   
2        2010.0          $2.901B           Asset             39.0        12.0   
3        2005.0          $3.472B        Acquired             56.0        13.0   
4        1991.0          $4.377B        Acquired             25.0         5.0   

   Portfolio Exits  
0              5.0  
1              NaN  
2              1.0  
3              1.0  
4              2.0

In [212]:
#removing the dollar sign from the  Valuation ($) , Total Raised ($) columns
df_copy['Valuation ($)'] = df_copy['Valuation ($)'].apply(lambda x: x.replace('$',''))
df_copy['Total Raised ($)'] = df_copy['Total Raised ($)'].astype(str)
df_copy['Total Raised ($)'] = df_copy['Total Raised ($)'].apply(lambda x: x.replace('$',''))
#we did not remove the $ from Total Raised ($) because it was taking care of when it was converted to floats 

In [213]:
df_copy.head()

Company Valuation ($) Date Joined        Country           City  \
0   Bytedance           140    4/7/2017          China        Beijing   
1      SpaceX         100.3   12/1/2012  United States      Hawthorne   
2      Stripe            95   1/23/2014  United States  San Francisco   
3      Klarna          45.6  12/12/2011         Sweden      Stockholm   
4  Epic Games            42  10/26/2018  United States           Cary   

                  Industry                                  Select Inverstors  \
0  Artificial intelligence  Sequoia Capital China, SIG Asia Investments, S...   
1                    Other  Founders Fund, Draper Fisher Jurvetson, Rothen...   
2                  Fintech        Khosla Ventures, LowercaseCapital, capitalG   
3                  Fintech  Institutional Venture Partners, Sequoia Capita...   
4                    Other              Tencent Holdings, KKR, Smash Ventures   

   Founded Year Total Raised ($) Financial Stage  Investors Count  Deal Terms  \
0        2012.0            7.44B             IPO             28.0         8.0   
1        2002.0           6.874B             NaN             29.0        12.0   
2        2010.0           2.901B           Asset             39.0        12.0   
3        2005.0           3.472B        Acquired             56.0        13.0   
4        1991.0           4.377B        Acquired             25.0         5.0   

   Portfolio Exits  
0              5.0  
1              NaN  
2              1.0  
3              1.0  
4              2.0

In [214]:
#function that takes the str number , decide if it's million or buillion, from the letter,and add the coresponding zeros to it 
def change_m_b(input_num): 
    num = str(input_num)
    for i in range(len(num)): 
        if num[i] == 'B':
            float_num = float(num[:i])
            float_num = float_num * 1000000000
            return int(float_num)
        elif num[i] == 'M':
            float_num = float(num[:i])
            float_num = float_num * 1000000
            return int(float_num)
        else :
            pass 

In [215]:
#testing if the conversion code wroks as expected or not 
print(f"original value : {df_copy['Total Raised ($)'][1032]}, after chaning : {change_m_b(df_copy['Total Raised ($)'][1032])}")

original value : 181.06M, after chaning : 181060000


In [216]:
#applying the code to the Total Raised columns
for i in range(df.shape[0]):
    df_copy['Total Raised ($)'][i] = change_m_b(df_copy['Total Raised ($)'][i])  

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [217]:
# now doing the same for the Valuation
def change_b(input_num):
    num = str(input_num)
    for i in range(len(num)): 
            float_num = float(num)
            float_num = float_num * 1000000000
            return int(float_num)

In [218]:
#applying the code to the Valuation column
for i in range(df.shape[0]):
    df_copy['Valuation ($)'][i] = change_b(df_copy['Valuation ($)'][i])  

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [219]:
#let's see if the df dataset is changed correctly or not compared to the original dataset data
df.head()


Company Valuation ($B) Date Joined        Country           City  \
0   Bytedance           $140    4/7/2017          China        Beijing   
1      SpaceX         $100.3   12/1/2012  United States      Hawthorne   
2      Stripe            $95   1/23/2014  United States  San Francisco   
3      Klarna          $45.6  12/12/2011         Sweden      Stockholm   
4  Epic Games            $42  10/26/2018  United States           Cary   

                  Industry                                  Select Inverstors  \
0  Artificial intelligence  Sequoia Capital China, SIG Asia Investments, S...   
1                    Other  Founders Fund, Draper Fisher Jurvetson, Rothen...   
2                  Fintech        Khosla Ventures, LowercaseCapital, capitalG   
3                  Fintech  Institutional Venture Partners, Sequoia Capita...   
4                    Other              Tencent Holdings, KKR, Smash Ventures   

   Founded Year Total Raised Financial Stage  Investors Count  Deal Terms  \
0        2012.0       $7.44B             IPO             28.0         8.0   
1        2002.0      $6.874B             NaN             29.0        12.0   
2        2010.0      $2.901B           Asset             39.0        12.0   
3        2005.0      $3.472B        Acquired             56.0        13.0   
4        1991.0      $4.377B        Acquired             25.0         5.0   

   Portfolio Exits  
0              5.0  
1              NaN  
2              1.0  
3              1.0  
4              2.0

In [220]:
df_copy.head()

Company Valuation ($) Date Joined        Country           City  \
0   Bytedance  140000000000    4/7/2017          China        Beijing   
1      SpaceX  100300000000   12/1/2012  United States      Hawthorne   
2      Stripe   95000000000   1/23/2014  United States  San Francisco   
3      Klarna   45600000000  12/12/2011         Sweden      Stockholm   
4  Epic Games   42000000000  10/26/2018  United States           Cary   

                  Industry                                  Select Inverstors  \
0  Artificial intelligence  Sequoia Capital China, SIG Asia Investments, S...   
1                    Other  Founders Fund, Draper Fisher Jurvetson, Rothen...   
2                  Fintech        Khosla Ventures, LowercaseCapital, capitalG   
3                  Fintech  Institutional Venture Partners, Sequoia Capita...   
4                    Other              Tencent Holdings, KKR, Smash Ventures   

   Founded Year Total Raised ($) Financial Stage  Investors Count  Deal Terms  \
0        2012.0       7440000000             IPO             28.0         8.0   
1        2002.0       6874000000             NaN             29.0        12.0   
2        2010.0       2901000000           Asset             39.0        12.0   
3        2005.0       3472000000        Acquired             56.0        13.0   
4        1991.0       4377000000        Acquired             25.0         5.0   

   Portfolio Exits  
0              5.0  
1              NaN  
2              1.0  
3              1.0  
4              2.0

In [221]:
#now we can assign df to df_copy
df = df_copy

# we don't need to change the None values with anything since the .describe() method will ignore the np.nan values when doing the calculations 

In [222]:
float_col_names = ['Valuation ($)','Founded Year','Total Raised ($)','Investors Count','Deal Terms','Portfolio Exits']
float_df = df[float_col_names]
float_df.dtypes 
# we can see that we have Valuation ($), and Total Raised ($) of type object, and the rest are type floats



Valuation ($)        object
Founded Year        float64
Total Raised ($)     object
Investors Count     float64
Deal Terms          float64
Portfolio Exits     float64
dtype: object

In [223]:
#cahinging Valuation ($), and Total Raised ($) to type floats using .astype() method
float_df['Valuation ($)'] = float_df['Valuation ($)'].astype(float)
float_df['Total Raised ($)'] = float_df['Total Raised ($)'].astype(float)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [224]:
#after checking we can see that all our columns now of type floats64
float_df.dtypes 

Valuation ($)       float64
Founded Year        float64
Total Raised ($)    float64
Investors Count     float64
Deal Terms          float64
Portfolio Exits     float64
dtype: object

In [225]:
describe_float_df = float_df.describe()
#creating a new column named range and having it's values max-min
describe_float_df.loc['range'] = describe_float_df.loc['max'] - describe_float_df.loc['min']
#the 50% quartile is just the median, so i changed it's name in the dataset 
describe_float_df.rename({'median':'50%'}, inplace=True)
#i am only interested in these field
ourfields = ['mean', '25%', '50%', '75%', 'range']
describe_float_df = describe_float_df.loc[ourfields]
#viewing the data set in transposed position 
describe_float_df.T


mean           25%           50%           75%  \
Valuation ($)     3.292083e+09  1.080000e+09  1.600000e+09  3.100000e+09   
Founded Year      2.012710e+03  2.011000e+03  2.014000e+03  2.016000e+03   
Total Raised ($)  5.780486e+08  2.285800e+08  3.810000e+08  6.400000e+08   
Investors Count   1.443243e+01  8.000000e+00  1.300000e+01  1.900000e+01   
Deal Terms        3.093254e+00  1.000000e+00  3.000000e+00  4.000000e+00   
Portfolio Exits   1.346939e+00  1.000000e+00  1.000000e+00  1.000000e+00   

                         range  
Valuation ($)     1.390000e+11  
Founded Year      1.020000e+02  
Total Raised ($)  7.439000e+09  
Investors Count   9.000000e+01  
Deal Terms        1.800000e+01  
Portfolio Exits   4.000000e+00

# adding the Industry column to the float_df and then group them by industry and plot hisotgram with the different industries 

In [226]:
float_df_copy = float_df.copy()
#creating a new column ['Industry']
float_df_copy['Industry'] = data['Industry']


In [227]:
float_df_copy_columns_list = float_df_copy.columns.tolist()
float_df_copy_columns_list

['Valuation ($)',
 'Founded Year',
 'Total Raised ($)',
 'Investors Count',
 'Deal Terms',
 'Portfolio Exits',
 'Industry']

In [228]:
float_df_copy_columns_list = ['Industry','Valuation ($)','Founded Year',
 'Total Raised ($)',
 'Investors Count',
 'Deal Terms',
 'Portfolio Exits']
float_df_copy = float_df_copy[float_df_copy_columns_list]
unique_float_df_copy = float_df_copy['Industry'].unique()
#unique_float_df_copy = pd.DataFrame(unique_float_df_copy)
unique_float_df_copy

array(['Artificial intelligence', 'Other', 'Fintech',
       'Internet software & services',
       'Supply chain, logistics, & delivery',
       'Data management & analytics',
       'Sequoia Capital, Thoma Bravo, Softbank',
       'E-commerce & direct-to-consumer', 'Edtech', 'Hardware',
       'Consumer & retail', 'Health', 'Auto & transportation', 'Finttech',
       'Cybersecurity', 'Mobile & telecommunications', 'Travel',
       'Kuang-Chi',
       'Tiger Global Management, Tiger Brokers, DCM Ventures',
       'Jungle Ventures, Accel, Venture Highway',
       'Artificial Intelligence',
       'Vision Plus Capital, GSR Ventures, ZhenFund',
       'Hopu Investment Management, Boyu Capital, DC Thomson Ventures',
       '500 Global, Rakuten Ventures, Golden Gate Ventures',
       'Sequoia Capital China, ING, Alibaba Entrepreneurs Fund',
       'Sequoia Capital China, Shunwei Capital Partners, Qualgro',
       'Dragonfly Captial, Qiming Venture Partners, DST Global',
       'SingTel Inn

In [229]:
#group them by industry 
#creating a new datset with the each industry having the total of the each parameters from float_df 
#and then showing them in a histogram 
#float_df_copy.plot.hist(bins = 25,alpha=0.5)


                                Valuation ($).      Total Raised ($)   

---



Artificial intelligence

---


Other


---


Internet software & services

In [230]:
#seting the index to the unique values in Industry, and the columns to the the float columns except for industry 
new_df = pd.DataFrame(0, index = unique_float_df_copy, columns= float_df_copy_columns_list )
del new_df['Industry']
new_df.head()



Valuation ($)  Founded Year  \
Artificial intelligence                          0             0   
Other                                            0             0   
Fintech                                          0             0   
Internet software & services                     0             0   
Supply chain, logistics, & delivery              0             0   

                                     Total Raised ($)  Investors Count  \
Artificial intelligence                             0                0   
Other                                               0                0   
Fintech                                             0                0   
Internet software & services                        0                0   
Supply chain, logistics, & delivery                 0                0   

                                     Deal Terms  Portfolio Exits  
Artificial intelligence                       0                0  
Other                                         0                0  
Fintech                                       0                0  
Internet software & services                  0                0  
Supply chain, logistics, & delivery           0                0

In [231]:

m = df[df['Industry'] == 'Artificial intelligence'].sum()
m


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  


Company              BytedanceFaireScale AIArgo AIGongAutomation An...
Valuation ($)                                             308389999999
Date Joined          4/7/201710/30/20198/5/20197/12/20198/12/20207/...
Country              ChinaUnited StatesUnited StatesUnited StatesUn...
City                 BeijingSan FranciscoSan FranciscoPittsburghPal...
Industry             Artificial intelligenceArtificial intelligence...
Select Inverstors    Sequoia Capital China, SIG Asia Investments, S...
Founded Year                                                  138894.0
Total Raised ($)                                           40882619999
Investors Count                                                 1115.0
Deal Terms                                                       209.0
Portfolio Exits                                                    9.0
dtype: object

In [232]:
'''creating two loops to go through the new_df.index, new_df.columns
m is going to be equal to the sum of all the values in the df dataset that has the value of i in it's ['Industry'] column and save the values in sum_df
and assigning the column j and row i of new_df to sum_df of row j 
'''
for i in new_df.index:
  sum_df = df[df['Industry'] == i].sum() 
  for j in new_df.columns:
    new_df[j][i] = sum_df[j]
new_df.head()

#since the Founded Year now makes no since we will change it with a function that seach for the range 
for i in new_df.index:
  sum_df = df[df['Founded Year'] == i].min() 

sum_df.head()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  


Company          NaN
Valuation ($)    NaN
Date Joined      NaN
Country          NaN
City             NaN
dtype: object

In [237]:
#this code will  seach the Founded Year column of df and find the min year and the max year, and will save it as tuple 
for i in new_df.index:
  min,max = int(df['Founded Year'].min()), int(df['Founded Year'].max())
  new_df['Founded Year'][i] = str(min) + '-' + str(max)

new_df.head()

Valuation ($) Founded Year  \
Artificial intelligence               308389999999    1919-2021   
Other                                 265200000000    1919-2021   
Fintech                               784820000000    1919-2021   
Internet software & services          556130000000    1919-2021   
Supply chain, logistics, & delivery   174759999999    1919-2021   

                                     Total Raised ($)  Investors Count  \
Artificial intelligence                   40882619999             1115   
Other                                     43665739999              724   
Fintech                                   97032519996             3207   
Internet software & services              73053259998             2736   
Supply chain, logistics, & delivery       41911720000              852   

                                     Deal Terms  Portfolio Exits  
Artificial intelligence                     209                9  
Other                                       141                6  
Fintech                                     601               20  
Internet software & services                616                6  
Supply chain, logistics, & delivery         191                1